### PR TITLE
feat: #170 - Emit kubernetes events

### PR DESCRIPTION
**Pull request by grzesuav**

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -18,10 +18,17 @@ package server
 
 import (
 	"fmt"
+	corev1 "k8s.io/api/core/v1"
```
Comment (grzesuav) [inline server/server.go:21]: need to check if this dependency is needed

File: server/server.go [hunk 2]
```diff
@@ -63,11 +79,22 @@ func Start(config *rest.Config, discoveryInterval, informerRelist time.Duration,
...
 
+	// Create EventBroadcaster, for creating EventRecorders
+	broadcaster := record.NewBroadcaster()
+
+	clientSet, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to init clientSet: %w", err)
+	}
+	// TODO check server API version and record to either new or old events API
+	broadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: clientSet.CoreV1().Events(metav1.NamespaceAll)})
```
Comment (grzesuav) [inline server/server.go:90]: TODO: when run on older cluster, there might be earlier version of events API, need to add mechanism to recognize that. Also, need to check in which version of events were added

**Comment by grzesuav** [comment]
Example output after running e2e tests :
```
 grzesiek@zenbook  ~  k get events --namespace default --sort-by='.lastTimestamp'                                                                                                                                                 ✔  20:32:24 
LAST SEEN   TYPE     REASON     OBJECT                                                 MESSAGE
2m35s       Normal   Started    compositecontroller/bluegreen-controller               Started controller: bluegreen-controller
2m35s       Normal   Starting   compositecontroller/bluegreen-controller               Starting controller: bluegreen-controller
2m9s        Normal   Stopping   compositecontroller/bluegreen-controller               Stopping controller: bluegreen-controller
2m9s        Normal   Stopped    compositecontroller/bluegreen-controller               Stopped controller: bluegreen-controller
2m9s        Normal   Started    decoratorcontroller/cluster-parent                     Started controller: cluster-parent
2m9s        Normal   Starting   decoratorcontroller/cluster-parent                     Starting controller: cluster-parent
2m3s        Normal   Stopped    decoratorcontroller/cluster-parent                     Stopped controller: cluster-parent
2m3s        Normal   Stopping   decoratorcontroller/cluster-parent                     Stopping controller: cluster-parent
112s        Normal   Started    compositecontroller/configmap-propagation-controller   Started controller: configmap-propagation-controller
112s        Normal   Starting   compositecontroller/configmap-propagation-controller   Starting controller: configmap-propagation-controller
105s        Normal   Stopping   compositecontroller/configmap-propagation-controller   Stopping controller: configmap-propagation-controller
105s        Normal   Stopped    compositecontroller/configmap-propagation-controller   Stopped controller: configmap-propagation-controller
105s        Normal   Starting   decoratorcontroller/crd-role-controller                Starting controller: crd-role-controller
105s        Normal   Started    decoratorcontroller/crd-role-controller                Started controller: crd-role-controller
102s        Normal   Stopping   decoratorcontroller/crd-role-controller                Stopping controller: crd-role-controller
102s        Normal   Stopped    decoratorcontroller/crd-role-controller                Stopped controller: crd-role-controller
81s         Normal   Started    compositecontroller/daemonjob-controller               Started controller: daemonjob-controller
81s         Normal   Starting   compositecontroller/daemonjob-controller               Starting controller: daemonjob-controller
36s         Normal   Stopping   compositecontroller/daemonjob-controller               Stopping controller: daemonjob-controller
36s         Normal   Stopped    compositecontroller/daemonjob-controller               Stopped controller: daemonjob-controller
15s         Normal   Started    compositecontroller/globalconfigmap-controller         Started controller: globalconfigmap-controller
15s         Normal   Starting   compositecontroller/globalconfigmap-controller         Starting controller: globalconfigmap-controller
9s          Normal   Stopping   compositecontroller/globalconfigmap-controller         Stopping controller: globalconfigmap-controller
9s          Normal   Stopped    compositecontroller/globalconfigmap-controller         Stopped controller: globalconfigmap-controller
```

File: controller/composite/controller.go [hunk 3]
```diff
@@ -67,13 +71,14 @@ type parentController struct {
...
-	numWorkers int
+	numWorkers    int
+	eventRecorder record.EventRecorder
 
 	finalizer *finalizer.Manager
 	customize customize.Manager
 }
 
-func newParentController(resources *dynamicdiscovery.ResourceMap, dynClient *dynamicclientset.Clientset, dynInformers *dynamicinformer.SharedInformerFactory, mcClient mcclientset.Interface, revisionLister mclisters.ControllerRevisionLister, cc *v1alpha1.CompositeController, numWorkers int) (pc *parentController, newErr error) {
+func newParentController(resources *dynamicdiscovery.ResourceMap, dynClient *dynamicclientset.Clientset, dynInformers *dynamicinformer.SharedInformerFactory, mcClient mcclientset.Interface, revisionLister mclisters.ControllerRevisionLister, cc *v1alpha1.CompositeController, numWorkers int, eventRecorder record.EventRecorder) (pc *parentController, newErr error) {
```
Comment (AmitKumarDas) [inline controller/composite/controller.go:81]: Do you think we can refactor this as following:

```go
// ParentControllerConfig is used to create new instances of
// ParentController
type ParentControllerConfig {
  Field1 int
  Field2 string
  Field3 string
}

type ParentController {
  field1 int
  field2 string
  field3 string
  field4 string
}

func NewParentController(config ParentControllerConfig) (*ParentController, error) {
  // some logic ...

  pc := &ParentController{
    field1: config.Field1,
    field2: config.Field2,
    field3: config.Field3,
  }
  
  // some other logic ...
  return pc, nil
}
```

Comment (AmitKumarDas) [reply controller/composite/controller.go:81]: Above pattern ensures a fixed signature. 
This reduces the impact of changes at caller code.

Comment (grzesuav) [reply controller/composite/controller.go:81]: I did that for `Config`,  but looking at 
```go
func newParentController(resources *dynamicdiscovery.ResourceMap, dynClient *dynamicclientset.Clientset, dynInformers *dynamicinformer.SharedInformerFactory, mcClient mcclientset.Interface, revisionLister mclisters.ControllerRevisionLister, cc *v1alpha1.CompositeController, numWorkers int, eventRecorder record.EventRecorder) (pc *parentController, newErr error) {
```
I do not see clear separation what we could consider `configuration` here, except `numWorkers int` - others are objects required for running.

I believe we can think of a bit of refactoring, as most of this it taken from `mc` - 
```go
pc, err := newParentController(mc.resources, mc.dynClient, mc.dynInformers, mc.mcClient, mc.revisionLister, cc, mc.numWorkers, mc.eventRecorder)
```


but would like to do it in separate pull request, to do not mix too many changes together

Comment (AmitKumarDas) [reply controller/composite/controller.go:81]: If not config then options is good naming.

Comment (grzesuav) [reply controller/composite/controller.go:81]: sure, however will do separate Pull Request, to not adding more changes here.

Comment (grzesuav) [reply controller/composite/controller.go:81]: follow up PR - in progress- https://github.com/metacontroller/metacontroller/pull/223

File: main.go [hunk 4]
```diff
@@ -76,7 +81,18 @@ func main() {
 	config.QPS = float32(*clientGoQPS)
 	config.Burst = *clientGoBurst
 
-	stopServer, err := server.Start(config, *discoveryInterval, *informerRelist, *workers)
+	options := options.Options{
```
Comment (AmitKumarDas) [inline main.go:84]: I felt placing Options within specific packages will make things more cohesive. 
However, we shall take a look at next PR.

**Review by AmitKumarDas — APPROVED** [review]

**Pull request merged**